File: src/pattern/iterator/java/BookShelfIterator.java
package pattern.iterator.java;

public class BookShelfIterator implements Iterator {
    private int index;
    private BookShelf bookShelf;

    public BookShelfIterator( BookShelf bookShelf){
        this.index = 0;
        this.bookShelf = bookShelf;
    }

    public boolean hasNext(){
        if (this.index < bookShelf.getLength())
            return true;
        else
            return false;
    }

    public Object next(){
        Book book = this.bookShelf.getBookAt(index);
        index++;

        return book;
    }
}

File: src/pattern/iterator/kotlin/BookShelfIterator.kt
package pattern.iterator.kotlin

class BookShelfIterator(var index: Int, var bookShelf: BookShelf): Iterator {

    override fun hasNext(): Boolean{
        return (index < bookShelf.getLength())
    }

    override fun next(): Any{
        val book = bookShelf.getBookAt(this.index)
        this.index++

        return book
    }
}